assets/js/upload.js

```diff
@@ -171,9 +171,121 @@ function setupImageUpload() {
     window.removeEventListener('beforeunload', beforeUnload);
   }
 
+  function createTagError(message) {
+    const buttonAfter = $('#tagsinput-save');
+    const errorElement = makeEl('span', { className: 'help-block tag-error' });
+
+    errorElement.innerText = message;
+    buttonAfter.insertAdjacentElement('beforebegin', errorElement);
+  }
+
+  function clearTagErrors() {
+    $$('.tag-error').forEach(el => el.remove());
+  }
+
+  const ratingsTags = ['safe', 'suggestive', 'questionable', 'explicit', 'semi-grimdark', 'grimdark', 'grotesque'];
+
+  // populate tag error helper bars as necessary
+  // return true if all checks pass
+  // return false if any check fails
+  function validateTags() {
+    const tagInput = $('textarea.js-taginput');
+
+    if (!tagInput) {
+      return true;
+    }
+
+    const tagsArr = tagInput.value.split(',').map(t => t.trim());
+
+    const errors = [];
+
+    let hasRating = false;
+    let hasSafe = false;
+    let hasOtherRating = false;
+
+    tagsArr.forEach(tag => {
+      if (ratingsTags.includes(tag)) {
+        hasRating = true;
+        if (tag === 'safe') {
+          hasSafe = true;
+        } else {
+          hasOtherRating = true;
+        }
+      }
+    });
+
+    if (!hasRating) {
+      errors.push('Tag input must contain at least one rating tag');
+    } else if (hasSafe && hasOtherRating) {
+      errors.push('Tag input may not contain any other rating if safe');
+    }
+
+    if (tagsArr.length < 3) {
+      errors.push('Tag input must contain at least 3 tags');
+    }
+
+    errors.forEach(msg => createTagError(msg));
+
+    return errors.length === 0; // true: valid if no errors
+  }
+
+  function enableUploadButton() {
+    const submitButton = $('.input--separate-top');
+    if (submitButton !== null) {
+      submitButton.disabled = false;
+      submitButton.innerText = 'Upload';
+    }
+  }
+
+  function disableUploadButton() {
+    const submitButton = $('.button.input--separate-top');
+    if (submitButton !== null) {
+      submitButton.disabled = true;
+      submitButton.innerText = 'Please wait...';
+    }
+
+    // delay is needed because Safari stops the submit if the button is immediately disabled
+    requestAnimationFrame(() => submitButton.setAttribute('disabled', 'disabled'));
+  }
+
+  function scrollToTags() {
+    const taginputEle = $('#taginput-fancy-tag_input');
+
+    if (!taginputEle) {
+      // default to scroll to top
+      window.scrollTo({ top: 0, left: 0 });
+    }
+
+    taginputEle.scrollIntoView();
```

Comment:
when reaching here, the tag block _must_ exist (because tag validation failed). Use `.js-tag-block` as the selector to handle the case of the plain editor being used. A separate function is maybe not even necessary anymore and could be inline at the call site.
```suggestion
    assertNotNull($('.js-tag-block')).scrollIntoView();
```

Reply:
It's not selecting that class, but '.js-taginput-input' is present.
Also, assertNotNull is undefined

Reply:
.js-taginput doesn't scroll into view but .js-taginput-input does

Reply:
Okay, that'll work then